p2p/libp2p/discovery/kadDhtDiscoverer2.go

```diff
@@ -0,0 +1,245 @@
+		randezVous:       randezVous,
+		initialPeersList: initialPeersList,
+	}
+}
+
+// Bootstrap will start the bootstrapping new peers process
+func (kdd *KadDhtDiscoverer2) Bootstrap() error {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+
+	if kdd.kadDHT != nil {
+		return p2p.ErrPeerDiscoveryProcessAlreadyStarted
+	}
+
+	if kdd.contextProvider == nil {
+		return p2p.ErrNilContextProvider
+	}
+
+	return kdd.startDHT()
+}
+
+// UpdateRandezVous change the randezVous string, and restart the discovery with the new protocols
+func (kdd *KadDhtDiscoverer2) UpdateRandezVous(s string) error {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+
+	if s == kdd.randezVous {
+		return nil
+	}
+
+	err := kdd.stopDHT()
+	if err != nil {
+		log.Debug("Error wile stopping kad-dht discovery, skip", "error", err)
+	}
+	kdd.randezVous = s
+	return kdd.startDHT()
+}
+
+func (kdd *KadDhtDiscoverer2) protocols() []protocol.ID {
+	return []protocol.ID{
+		protocol.ID(fmt.Sprintf("%s/erd_%s", opts.ProtocolDHT, kdd.randezVous)),
+		protocol.ID(fmt.Sprintf("%s/erd", opts.ProtocolDHT)),
+		//TODO: to be removed once the seed is updated
+		opts.ProtocolDHT,
+	}
+}
+
+func (kdd *KadDhtDiscoverer2) startDHT() error {
+	ctx := kdd.contextProvider.Context()
+	h := kdd.contextProvider.Host()
+
+	ctxrun, cancel := context.WithCancel(ctx)
+	hd, err := NewHostDecorator(h, ctxrun, 3, time.Second)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	kademliaDHT, err := dht.New(ctx, hd, opts.Protocols(kdd.protocols()...))
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	go kdd.connectToInitialAndBootstrap(ctxrun)
+
+	kdd.kadDHT = kademliaDHT
+	kdd.refreshCancel = cancel
+	return nil
+}
+
+func (kdd *KadDhtDiscoverer2) stopDHT() error {
+	if kdd.refreshCancel == nil {
+		return nil
+	}
+
+	kdd.refreshCancel()
+	kdd.refreshCancel = nil
+
+	h := kdd.contextProvider.Host()
+
+	for _, p := range kdd.protocols() {
+		h.RemoveStreamHandler(p)
+	}
+
+	err := kdd.kadDHT.Close()
+
+	kdd.kadDHT = nil
+
+	return err
+}
+
+func (kdd *KadDhtDiscoverer2) connectToInitialAndBootstrap(ctx context.Context) {
+	chanStartBootstrap := kdd.connectToOnePeerFromInitialPeersList(
+		kdd.refreshInterval,
+		kdd.initialPeersList)
+
+	go func() {
+		<-chanStartBootstrap
+
+		go func() {
+			for {
+				kdd.mutKadDht.RLock()
+				kadDht := kdd.kadDHT
+				kdd.mutKadDht.RUnlock()
+
+				var err = error(nil)
+				if kadDht != nil {
+					err = kadDht.Bootstrap(ctx)
+				}
+				if err == kbucket.ErrLookupFailure {
+					<-kdd.ReconnectToNetwork()
+				}
+				select {
+				case <-time.After(kdd.refreshInterval):
+				case <-ctx.Done():
+					return
+				}
+			}
+
+		}()
+	}()
+}
+
+func (kdd *KadDhtDiscoverer2) connectToOnePeerFromInitialPeersList(
+	intervalBetweenAttempts time.Duration,
+	initialPeersList []string) <-chan struct{} {
+
+	h := kdd.contextProvider.Host()
+	ctx := kdd.contextProvider.Context()
+
+	chanDone := make(chan struct{}, 1)
+
+	if initialPeersList == nil {
```

Comment:
this check is not needed - it is resolved by the next len(initialPeerList)

Reply:
Agree. Old code :D